dali/operators/generic/slice/subscript.h

```diff
@@ -0,0 +1,325 @@
+    const auto &input_shape = input.shape();
+    outputs.resize(1);
+    outputs[0].type = input.type();
+    GetRanges(ws, input_shape);
+    ProcessShapes(outputs[0].shape, input_shape);
+    return true;
+  }
+
+  void GetRanges(const workspace_t<Backend> &ws, const TensorListShape<> &in_shape) {
+    int nsub = subscripts_.size();
+    int ndim = in_shape.sample_dim();
+    DALI_ENFORCE(ndim >= nsub, "Too many indices.");
+    int nsamples = in_shape.num_samples();
+    lo_.resize(nsamples, ndim);
+    hi_ = in_shape;
+    step_.resize(nsamples, ndim);
+    for (auto &s : step_.shapes)
+      s = 1;
+    shape_ = in_shape;
+
+    for (int d = 0; d < nsub; d++) {
+      SubscriptInfo &s = subscripts_[d];
+      s.Load(ws, nsamples);
+      GetRange(s, d, in_shape);
+    }
+  }
+
+  void GetRange(SubscriptInfo &s, int d, const TensorListShape<> &in_shape) {
+    int nsamples = in_shape.num_samples();
+
+    if (s.IsIndex()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t at = s.at.values[i];
+        int64_t idx = at < 0 ? in_extent + at : at;
+        DALI_ENFORCE(idx >= 0 && idx < in_extent, make_string("Index ", idx, " is out of range "
+          "for axis ", d, " of length ", in_extent, "\n"
+          "Detected while processing sample #", i, " of shape (", in_shape[i], ")"));
+        lo_.tensor_shape_span(i)[d] = idx;
+        hi_.tensor_shape_span(i)[d] = idx + 1;
+        shape_.tensor_shape_span(i)[d] = 1;
+      }
+    }
+    if (s.IsRange()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t lo = s.lo.IsDefined() ? s.lo.values[i] : 0;
+        int64_t hi = s.hi.IsDefined() ? s.hi.values[i] : in_extent;
+        int64_t step = s.step.IsDefined() ? s.step.values[i] : 1;
+        DALI_ENFORCE(step == 1, "Indexing with non-unit step is not implemented");
+        if (lo < 0) lo += in_extent;
+        if (hi < 0) hi += in_extent;
+        lo = clamp(lo, 0_i64, in_extent);
+        hi = clamp(hi, 0_i64, in_extent);
+        lo_.tensor_shape_span(i)[d] = lo;
+        hi_.tensor_shape_span(i)[d] = hi;
+        step_.tensor_shape_span(i)[d] = step;
+
+        int64_t out_extent = step > 0 ? div_ceil(hi - lo,  step)
+                                      : div_ceil(lo - hi, -step);
+        if (out_extent < 0)
+          out_extent = 0;
+        shape_.tensor_shape_span(i)[d] = out_extent;
+      }
+    }
+  }
+
+  void ProcessShapes(TensorListShape<> &out_shape, const TensorListShape<> &in_shape) {
+    int in_dims = in_shape.sample_dim();
+    int nsub = subscripts_.size();
+
+    out_dim_map_.clear();
+    dim_groups_.clear();
+
+    int group_start = 0;
+    int d = 0;
+    for (; d < nsub; d++) {
+      if (subscripts_[d].IsDefined()) {
+        if (d != group_start)
+          dim_groups_.push_back({ group_start, d - group_start });
```

Comment:
This contains groups of dims where subscripts_ is not defined, right?

Reply:
I can rename to `collapsed_dims_` or sth.

Reply:
Sounds better